### PR TITLE
WIP: Initial FollowIR integration

### PR DESCRIPTION
**Pull request by orionw**
Adds a new abstract task to implement retrieval with instructions. Notes on code regarding the reasoning behind the new classes.

This is work in progress, I expect to get feedback on this one before making a final PR as I expect there will be some larger changes :) 

NOTE: this does not have rerankers, just instruction following

### CODE REVIEW

File: mteb/abstasks/AbsTaskInstructionRetrieval.py [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+from datasets import load_dataset, Value, Features
+
+from ..evaluation.evaluators import InstructionRetrievalEvaluator
+from ..evaluation.evaluators import utils
+from .AbsTask import AbsTask
+
+logger = logging.getLogger(__name__)
+
+# Adapted from https://github.com/beir-cellar/beir/blob/f062f038c4bfd19a8ca942a9910b1e0d218759d4/beir/datasets/data_loader_hf.py#L10
+class HFDataLoader:
```
Comment (orionw) [inline mteb/abstasks/AbsTaskInstructionRetrieval.py:18]: I had to modify this class because it now comes with a bunch of extra fields, such as the instructions, the changed_instructions, the top_ranked, etc.

File: mteb/abstasks/AbsTaskInstructionRetrieval.py [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+        features = Features({'query-id': Value('string'), 'corpus-id': Value('string'), 'score': Value('float')})
+        qrels_ds = qrels_ds.cast(features)
+
+        if changed:
+            self.changed_qrels = qrels_ds
+        else:
+            self.og_qrels = qrels_ds
+
+
+class AbsTaskInstructionRetrieval(AbsTask):
```
Comment (orionw) [inline mteb/abstasks/AbsTaskInstructionRetrieval.py:161]: This task was needed as the evaluation requires evaluation twice over the queries, with two different instructions. 

It's modified to run with both sets of instructions and then compute the compared scores.

File: mteb/abstasks/AbsTaskInstructionRetrieval.py [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+                }
+
+        return changed_scores
+
+    def _evaluate_monolingual(self, retriever, corpus, queries, relevant_docs, qrels_name: str, instructions, top_ranked, lang=None, **kwargs):
+        start_time = time()
+
+        # do the results by query and relevant docs only
+        all_results = []
+        for query_id in tqdm.tqdm(list(queries.keys()), leave=True):
```
Comment (orionw) [inline mteb/abstasks/AbsTaskInstructionRetrieval.py:262]: Unlike standard retrieval, this is just re-ranking the top N, so we do it query by query instead of all at once

File: mteb/evaluation/evaluators/InstructionRetrievalEvaluator.py [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+import torch
+from collections import defaultdict
+
+from .Evaluator import Evaluator
+from .utils import cos_sim, dot_score, mrr, recall_cap, hole, top_k_accuracy
+
+logger = logging.getLogger(__name__)
+
+# Adapted from https://github.com/beir-cellar/beir/blob/f062f038c4bfd19a8ca942a9910b1e0d218759d4/beir/retrieval/search/dense/exact_search.py#L12
+class DenseRetrievalExactSearch:
```
Comment (orionw) [inline mteb/evaluation/evaluators/InstructionRetrievalEvaluator.py:19]: Unfortunately, I had to modify this as well --- we need some way of caching the document embeddings.  I could move this class to AbstractRetrieval with the new additions?

File: mteb/evaluation/evaluators/InstructionRetrievalEvaluator.py [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+
+
+        for qid in results:
+            for doc_id, score in results[qid]:
+                self.results[qid][doc_id] = score
+        
+        return self.results
+
+
+class DRESModel:
```
Comment (orionw) [inline mteb/evaluation/evaluators/InstructionRetrievalEvaluator.py:108]: Same as with the exact search model, I did this because we cache the document embeddings.  I could move this into AbstractRetrieval though?

File: mteb/evaluation/evaluators/utils.py [hunk 6]
```diff
@@ -189,3 +190,54 @@ def top_k_accuracy(
...
+    sorted_by_score = sorted(tuple_of_id_score, key=lambda x: x[1], reverse=True)
+    # return the rank of the doc_id, if not found return -1
+    for i, (id, score) in enumerate(sorted_by_score):
+        if id == doc_id:
+            return i + 1, score
+        
+    return len(sorted_by_score) + 1, 0
+
+
+def evaluate_change(original_run, new_run, changed_qrels):
```
Comment (orionw) [inline mteb/evaluation/evaluators/utils.py:207]: This is the new p-MRR metric

File: mteb/tasks/InstructionReranking/CoreReranking.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
+from ...abstasks.AbsTaskInstructionRetrieval import AbsTaskInstructionRetrieval
+
+
+class NewsReranking(AbsTaskInstructionRetrieval):
```
Comment (orionw) [inline mteb/tasks/InstructionReranking/CoreReranking.py:4]: I think these task descriptions are older now, but I can replace them

File: mteb/abstasks/AbsTaskInstructionRetrieval.py [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+logger = logging.getLogger(__name__)
+
+# Adapted from https://github.com/beir-cellar/beir/blob/f062f038c4bfd19a8ca942a9910b1e0d218759d4/beir/datasets/data_loader_hf.py#L10
+class HFDataLoader:    
+    def __init__(self, hf_repo: str = None, hf_repo_qrels: str = None, data_folder: str = None, prefix: str = None, corpus_file: str = "corpus.jsonl", query_file: str = "queries.jsonl", 
+                 qrels_folder: str = "qrels", qrels_file: str = "", streaming: bool = False, keep_in_memory: bool = False):
+        self.corpus = {}
+        self.queries = {}
+        self.qrels = {}
+        self.og_instructions = {}
```
Comment (orionw) [inline mteb/abstasks/AbsTaskInstructionRetrieval.py:24]: The instructions have to be in this file because they come from the underlying data and are not model specific, unlike the prefixes that models currently use as instructions that are model-specific.

File: mteb/abstasks/AbsTaskInstructionRetrieval.py [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+            short_instructions = {query['text']: query['short_query'] for query in queries}
+            queries = {query['id']: query['text'] for query in queries}
+            assert len(top_ranked) == len(queries), f"Top ranked not loaded properly! Expected {len(self.queries)} but got {len(self.top_ranked)}."
+            corpus = {doc['id']: {'title': doc['title'] , 'text': doc['text']} for doc in corpus}
+            self.corpus[split], self.queries[split], self.og_relevant_docs[split], self.changed_relevant_docs[split] = corpus, queries, og_qrels, changed_qrels
+            self.changed_instructions[split], self.og_instructions[split] = changed_instructions, og_instructions
+            self.top_ranked[split] = top_ranked
+            self.keywords[split], self.short_instructions[split] = keywords, short_instructions
+
+        self.data_loaded = True
```
Comment (Muennighoff) [inline mteb/abstasks/AbsTaskInstructionRetrieval.py:197]: I think `self.description` has changed to `self.metadata_dict`

File: mteb/tasks/InstructionReranking/CoreReranking.py (hunk 7 above)
Comment (Muennighoff) [reply mteb/tasks/InstructionReranking/CoreReranking.py:4]: Yeah they have been updated cc @KennethEnevoldsen @MartinBernstorff

File: mteb/tasks/__init__.py [hunk 10]
```diff
@@ -8,3 +8,4 @@
 from .Retrieval import *
 from .STS import *
 from .Summarization import *
+from .InstructionReranking import *
```
Comment (Muennighoff) [inline mteb/tasks/__init__.py:11]: I think the folder structure has also changed so there should probably be an English folder cc @KennethEnevoldsen ?

Also, why `InstructionReranking` and not `InstructionRetrieval` or `FollowIR`?

File: mteb/__init__.py [hunk 11]
```diff
@@ -77,6 +77,14 @@
     "TwitterURLCorpus",
 ]
 
+
+MTEB_INSTRUCTION_FOLLOWING = [
```
Comment (Muennighoff) [inline mteb/__init__.py:81]: Maybe `MTEB_RETRIEVAL_INSTRUCTIONS` or `MTEB_RETRIEVAL_FOLLOWIR`?

**Review by Muennighoff — COMMENTED** [review]
Really excited about this integration, great work!

File: mteb/evaluation/evaluators/InstructionRetrievalEvaluator.py [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+        queries = [queries[qid] for qid in queries]
+        query_embeddings = self.model.encode_queries(
+            queries, batch_size=self.batch_size, show_progress_bar=self.show_progress_bar, convert_to_tensor=self.convert_to_tensor, instructions=instructions)
+          
+        logger.info("Sorting Corpus by document length (Longest first)...")
+
+        corpus_ids = sorted(corpus, key=lambda k: len(corpus[k].get("title", "") + corpus[k].get("text", "")), reverse=True)
+        corpus = [corpus[cid] for cid in corpus_ids]
+
+        logger.info("Encoding Corpus in batches... Warning: This might take a while!")
```
Comment (KennethEnevoldsen) [inline mteb/evaluation/evaluators/InstructionRetrievalEvaluator.py:65]:
```suggestion
        logger.info("Encoding Corpus in batches. This might take a while!")
```
Warning seems to suggest that something might be wrong.

File: mteb/evaluation/evaluators/InstructionRetrievalEvaluator.py [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+        # Runs semantic search against the corpus embeddings
+        # Returns a ranked list with the corpus ids
+        if score_function not in self.score_functions:
+            raise ValueError("score function: {} must be either (cos_sim) for cosine similarity or (dot) for dot product".format(score_function))
+            
+        logger.info("Encoding Queries...")
+        query_ids = list(queries.keys())
+        self.results = {qid: {} for qid in query_ids}
+        queries = [queries[qid] for qid in queries]
+        query_embeddings = self.model.encode_queries(
```
Comment (KennethEnevoldsen) [inline mteb/evaluation/evaluators/InstructionRetrievalEvaluator.py:57]: Should we document somewhere that it takes instructions?

**Review by KennethEnevoldsen — COMMENTED** [review]
Thanks for the addition! I have added a bunch of comments, some of them are to understand the code better.

File: mteb/abstasks/AbsTaskInstructionRetrieval.py (hunk 2 above)
Comment (orionw) [reply mteb/abstasks/AbsTaskInstructionRetrieval.py:161]: Good ideas, thanks @KennethEnevoldsen!

File: mteb/tasks/__init__.py (hunk 10 above)
Comment (orionw) [reply mteb/tasks/__init__.py:11]: This is a good question.  We could do InstructionRetrieval but in the case of FollowIR it's actually reranking not retrieval (1000 docs per query).  

There's another paper concurrent with mine that took an automated approach to data creation from [KAIST](https://github.com/kaistAI/InstructIR/tree/main).  I think I will see if they're interested in merging into MTEB also -- people always like to evaluate on more than one dataset anyways.  For that reason, I was thinking of keeping it more broad, so that it wasn't just FollowIR.

I think InstructionalRetrieval is more broad (which I like so that people don't think it can only be done for reranking) but InstructionReranking is technically more accurate. TBH I'm happy with either, both are good.  Do you have a preference @Muennighoff?

File: mteb/abstasks/AbsTaskInstructionRetrieval.py [hunk 14]
```diff
@@ -0,0 +1,312 @@
...
+        if not os.path.exists(fIn):
+            raise ValueError("File {} not present! Please provide accurate file.".format(fIn))
+        
+        if not fIn.endswith(ext):
+            raise ValueError("File {} must be present with extension {}".format(fIn, ext))
+
+    def load(self, split="test") -> Tuple[Dict[str, Dict[str, str]], Dict[str, str], Dict[str, Dict[str, int]], Dict[str, Dict[str, int]]]:
+        if not self.hf_repo:
+            self.og_qrels_file = os.path.join(self.qrels_folder + "_og", split + ".tsv")
+            self.changed_qrels_file = os.path.join(self.qrels_folder + "_changed", split + ".tsv")
```
[the remark this answers is not in the document]
Comment (orionw) [reply mteb/abstasks/AbsTaskInstructionRetrieval.py:58]: Ah, I personally dislike the fact that they are separate objects but that's just a personal preference. If the standard in this library is Path, I can switch!

File: mteb/tasks/InstructionReranking/NewsReranking.py [hunk 15]
```diff
@@ -0,0 +1,18 @@
+from ...abstasks.AbsTaskInstructionRetrieval import AbsTaskInstructionRetrieval
+
+
+class NewsReranking(AbsTaskInstructionRetrieval):
```
[the remark this answers is not in the document]
Comment (orionw) [reply mteb/tasks/InstructionReranking/NewsReranking.py:4]: Ah great point, I forgot about adding tests.  Thanks!!

File: mteb/abstasks/AbsTaskInstructionRetrieval.py [hunk 16]
```diff
@@ -0,0 +1,312 @@
...
+                for qid in list(results.keys()):
+                    doc_ids = set(sorted(results[qid], key=lambda x: results[qid][x], reverse=True)[:top_k])
+                    results[qid] = {k: v for k, v in results[qid].items() if k in doc_ids}
+            if lang is None:
+                qrels_save_path = f"{output_folder}/{self.description['name']}_qrels_{qrels_name}.json"
+            else:
+                qrels_save_path = f"{output_folder}/{self.description['name']}_{lang}_qrels_{qrels_name}.json"
+            
+            with open(qrels_save_path, "w") as f:
+                json.dump(results, f)
```
[the remark this answers is not in the document]
Comment (orionw) [reply mteb/abstasks/AbsTaskInstructionRetrieval.py:290]: Good catch, this is leftover from my debugging and I missed it.  Thanks!

File: mteb/evaluation/evaluators/InstructionRetrievalEvaluator.py (hunk 4 above)
Comment (KennethEnevoldsen) [reply mteb/evaluation/evaluators/InstructionRetrievalEvaluator.py:19]: Ah sorry - meant add to existing

**Comment by orionw** [comment]
Thanks for the feedback @KennethEnevoldsen and @Muennighoff! I made the changes starting from a fresh copy of MTEB to avoid all the merge conflicts and be sure I had the newest version. I'll make a real PR now instead of this draft PR.
